### PR TITLE
Create README.md

### PR DESCRIPTION
**Pull request by jayesh-parmar**
Create  Readme File  and add How To install project stap by stap

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,60 +7,23 @@
 <a href="https://packagist.org/packages/laravel/framework"><img src="https://img.shields.io/packagist/l/laravel/framework" alt="License"></a>
 </p>
 
-## About Laravel
+##  Laravel + Tailwind Css Admin Project with (Admin and Front-end)
+- login user
```
Comment (dineshram7660) [inline README.md]:
```suggestion
##  Laravel + Tailwind Css Admin Project 
- login user 
```

File: README.md [hunk 2]
```diff
@@ -7,60 +7,23 @@
 <a href="https://packagist.org/packages/laravel/framework"><img src="https://img.shields.io/packagist/l/laravel/framework" alt="License"></a>
 </p>
 
-## About Laravel
+##  Laravel + Tailwind Css Admin Project with (Admin and Front-end)
+- login user 
+- logout user
```
Comment (dineshram7660) [inline README.md:23]: a work kare che joy lejo no hoy to new pr api dejo logout mate

File: README.md [hunk 3]
```diff
@@ -7,60 +7,23 @@
...
+- open .env and update DB_DATABASE (database details)
+- run : `composer install`
+- run : `php artisan key:generate`
+- run : `php artisan migrate:fresh --seed`
+- run : `php artisan serve`
+- 
+## For Login
+- #### Admin
+- email: admin@gmail.com.com
+- password : admin123
```
Comment (dineshram7660) [inline README.md]:
```suggestion
- email: admin@gmail.com
- password : admin123
```

**Review by dineshram7660 — CHANGES_REQUESTED** [review]

**Comment by dineshram7660** [comment]
# Retail ERP

### Documentation
- https://www.notion.so/ERP-Development-Specifications-5a45ec1eaf4945a39c16cc5b777c6b12
We use Notion to record all tasks and track progress. Bugs/features/updates all should be recorded in Notion.

1) If you're discussing something with a team member or decision makers, record the important notes/ideas in a Notion task.
2) If updates in your code affects other files,
    i) Either make the necessary changes right away when it's needs to be deployed together or short in nature
    ii) Add a Notion task with details and link it with the PR both ways.

TLDR: There is no harm in creating lots of Notion tasks even if they need to be archived later on. But a lot of potential loss of missing important stuff if one is lazy or over confident on memory.

### Technical

The information about the development process and reference details are recorded in [this file](Technical.md).

### Important notes for developers

- We are using OCI High availability MySQL instance in production. [It requires](https://docs.oracle.com/en-us/iaas/mysql-database/doc/prerequisites1.html#GUID-CC4B1DEC-9C52-4387-83CB-188F0454BCFA) that all the tables in the database must have primary keys. The pivot tables in the Laravel app do not require a primary key but we are adding `invisible_id` as INVISIBLE column to support the OCI requirements. Please add `invisible_id` as primary key in new pivot tables.

```
$table->id('invisible_id')->invisible();
```

### Requirements

- PHP 8.2
- Other [Laravel requirements](https://laravel.com/docs/9.x/deployment#server-requirements)

### Installation

- Clone the repo:
`git clone [REPO_URL] [DIRECTORY_NAME]`

- Create `.env` file from the example file:
`php -r "file_exists('.env') || copy('.env.example', '.env');"`

- Setup .env variables (Especially SANCTUM_STATEFUL_DOMAINS should be set. The value should be the domain name without 'http://' and a trailing slash. And `LIST_OF_WEB_APP_URLS_AND_KEYS` should be updated in all the sites.)

- Install the dependencies: `composer install`

- Generate Key: `php artisan key:generate`

- DB migrate: `php artisan migrate`

- Public images: `php artisan storage:link`

- Set the `LOG_CHANNEL` to `production_stack` and specify the `LOG_SLACK_WEBHOOK_URL` if you wish to receive log entries in Slack.

- NPM: `npm install`

- `npm run dev` and start developing...

- Note: For the production just run the `npm run build` command

- Install [redis](https://www.digitalocean.com/community/tutorials/how-to-install-and-secure-redis-on-ubuntu-22-04) and [phpredis](https://github.com/phpredis/phpredis):
    ```shell
      sudo apt install -y redis-server
      sudo nano /etc/redis/redis.conf
   ```
    - Configure redis as follows:

    ```editorconfig
    . . .

    # If you run Redis from upstart or systemd, Redis can interact with your
    # supervision tree. Options:
    #   supervised no      - no supervision interaction
    #   supervised upstart - signal upstart by putting Redis into SIGSTOP mode
    #   supervised systemd - signal systemd by writing READY=1 to $NOTIFY_SOCKET
    #   supervised auto    - detect upstart or systemd method based on
    #                        UPSTART_JOB or NOTIFY_SOCKET environment variables
    # Note: these supervision methods only signal "process is ready."
    #       They do not enable continuous liveness pings back to your supervisor.
    supervised systemd

    . . .
    ```
    - Restart `redis`
    ```shell
    sudo systemctl restart redis.service
    ```

    - Test `redis`
    ```shell
    redis-cli
    ```
    - In the prompt that follows, test connectivity with the `ping` command and you should get `pong` response.

- Install [pickle](https://github.com/FriendsOfPHP/pickle):
```shell
wget https://github.com/FriendsOfPHP/pickle/releases/latest/download/pickle.phar
```
- Now install `phpredis` and use all the default settings.
```shell
sudo php pickle.phar install redis
```
- Setup Horizon(Add Daemon Process), and set .env variables accordingly.

- To allow to upload big size files run below commands from terminal.
```shell
    sudo sed -i "s/upload_max_filesize = .*/upload_max_filesize = 20M/" /etc/php/8.2/fpm/php.ini # Default: 2MB
    sudo sed -i "s/upload_max_filesize = .*/upload_max_filesize = 20M/" /etc/php/8.2/cli/php.ini # Default: 2MB
    sudo sed -i "s/post_max_size = .*/post_max_size = 30M/" /etc/php/8.2/fpm/php.ini # Default: 8MB This value must be larger than upload_max_filesize and lower than memory_limit
    sudo sed -i "s/post_max_size = .*/post_max_size = 30M/" /etc/php/8.2/cli/php.ini # Default: 8MB This value must be larger than upload_max_filesize and lower than memory_limit
    sudo sed -i -e 's/memory_limit = .*/memory_limit = 4096M/g' /etc/php/8.2/cli/php.ini # Default: 128MB
    sudo sed -i -e 's/memory_limit = .*/memory_limit = 4096M/g' /etc/php/8.2/fpm/php.ini # Default: 128MB
```

- If your any of the request will take longer time to execute operation run below commands to increase the execution time.
```shell
    sudo sed -i "s/max_execution_time = .*/max_execution_time = 1200/" /etc/php/8.2/fpm/php.ini # Default: 30 seconds
    sudo sed -i "s/max_execution_time = .*/max_execution_time = 1200/" /etc/php/8.2/cli/php.ini # Default: 30 seconds

    sudo sed -i "s/request_terminate_timeout = .*/request_terminate_timeout = 1200/" /etc/php/8.2/fpm/pool.d/www.conf # Default: Off (0)

    test -f /etc/nginx/conf.d/timeout.conf && echo "timeout.conf file already exists." || (sudo touch /etc/nginx/conf.d/timeout.conf && echo 'fastcgi_read_timeout 1200;' | sudo tee --append /etc/nginx/conf.d/timeout.conf) # Create timeout configuration file if not exists

    # To change the value of fastcgi_read_timeout. But, If you are setting up for the first time and you have fired above command this command is not required.
    sudo sed -i "s/fastcgi_read_timeout .*/fastcgi_read_timeout 1200;/" /etc/nginx/conf.d/timeout.conf # Default: 60 seconds
```

- We are using [Laravel Passport](https://laravel.com/docs/10.x/passport) for app authentication. Please run the following command to generate the keys required by the underlying library: `php artisan passport:keys`

- For the backward compatibility we are allowing personal access token for the `Member Application` and `POS Application` API's. For that you need to run `php artisan passport:client --personal`

- We need to increase the `client_max_body_size` of nginx to allow file uploads. Please add the following line in the the `/etc/nginx/nginx.conf` file's http directive:
    `client_max_body_size 50M;`

- We are using [Browsershot](https://github.com/spatie/browsershot) to generate PDF files. You may use the following commands to install the required dependencies. It also install NodeJS. You may skip first line and `nodejs` install command if it is already installed.
    ```
    sudo apt-get install -y nodejs gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgbm1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget libgbm-dev
    ```

- Timezone changed to MYT:
    - Forge > Server > Meta > Timezone (Running this command updates the server timezone. No need to fire the next command when forge is available)
    - (If you have done the change via forge, no need to follow this line) Server level - `sudo timedatectl set-timezone Asia/Kuala_Lumpur` (Command to check current timezone - `timedatectl`)
    - Laravel - application timezone is set to `Asia/Kuala_Lumpur` in the config file.
    - MySQL and PHP timezone depends on the server timezone. Just reload PHP FPM & MySQL services.
    - If there is a separate server for MySQL, you need to set the server timezone accordingly.

- Please test the PDF generation functionality once after setting up the server. Here are some of the possible reasons for issues:
    - [Custom Node and NPM](https://spatie.be/docs/browsershot/v2/requirements#content-custom-node-and-npm-binaries) path is needed when the server binaries are not in the default location. Check by running `which node` and `which npm` commands.
    - The directory where the PDF files are to be stored is not created. Puppeteer does not create the directory if it does not exist.
    - Global installation of puppeteer is mostly not needed. Local installation works.

- [Prevent main branch direct pushes](#prevent-main-branch-direct-pushes)

- Setup [Scheduler](https://laravel.com/docs/9.x/scheduling#main-content) to run every minute. We are generating the member's birthday vouchers every day via cronjob.

- Seed data using **one** of the following:
    - For production/local, run `php artisan db:seed --class=StaticDataSeeder` to seed static data required to run the system.
    - DB Seed: `php artisan db:seed` (Run if you need auto-generated sample data)
    - For seeding data for mobile POS API, run `php artisan db:seed --class=MobileApiSeeder` to seed 10k records.

- A record for Super admin need to be added manually to the database table `super_admins`.
    - For development: run `php artisan db:seed`. username: `super_admin`, password: `123456`
    - For production: Add a custom username. And password needs hashed by using the `bcrypt()` helper before saving it.

- When using Envoyer:
    - Set the deployment script in forge to blank as deployment happen from Envoyer only.
    - Edit the nginx config and change the root to include the 'current' release.
    - Need to add a 'Managed Linked Folder' for linking the `public/storage` to the `storage/app/public` directory after each deployment.
    - **Important**: We cannot create the folders inside the `storage` directory using the .gitignore file because Envoyer ignores that directory and just symlinks it to the `storage` directory at the root of the project. Either we need to create those directories manually there or we should use composer scripts to create the folders, if required, after each install command.


### Things to test
- Add or edit functionality of any module
- File uploads
- Import records to test Horizon + Queued jobs
- Emails
- File export
- PDF generation in Stock Transfer
- POS connection
- Automated backups
- Logs, Slack notifications, and Sentry

### Identification of the site by POS

Cashier needs to enter the configuration key before logging in. And POS makes an API call to one of the sites to fetch the URL of the site associated with the specified configuration key. Currently, we are adding all the keys to all the sites but long term we will take this to a separate site.


### Automated Tests

All the backend code has to be covered with automated tests. We are using [Pest](https://pestphp.com) to run automated PHP tests.

Run by `./vendor/bin/pest`

IMP: We are following domain driven coding approach for easier management of various modules and faster tests.

### Continuous Integration (CI) processes

[GitHub Actions](https://github.com/features/actions) are used to automate processes:
- [Pest](https://pestphp.com) for test runs
- [Rector](https://github.com/rectorphp/rector) for automated refactoring
- [ECS](https://github.com/symplify/easy-coding-standard) for automated coding standard fixes
- [Larastan](https://github.com/nunomaduro/larastan) for static analysis checks
- [ESLint](https://eslint.org/) for Javascript code quality checks


### Frontend assets setup

// TODO
- How vite is used to manage assets for multiple panels
- How we remove extra css using purgecss


### PHP version upgrade steps

1. Server level: Install new versin of PHP on the server from Forge UI. And set default to it on CLI as well as FPM.
1. Server level: Check if PHP redis extension needs to be updated.
1. Server level: memory limit, max upload file size, etc. config changes as per the README.
1. Server level: Install required PHP extensions for the application to run. (Mostly forge takes care of this.)
1. Site level: Update the PHP version from the Meta tab of the Forge UI.
1. (Optional) Site level: Update nginx site config to handle requests using the new php-fpm version
1. Server level: Reload nginx
1. Server level: Update the Scheduler and Daemons paths to reflect new PHP version.
1. Envoyer Site Level: Update Server -> Select new PHP Version
1. Deploy new code from Envoyer


### Prevent main branch direct pushes

1. Open terminal (not inside VS Code) and cd into the project directory

2. `touch .git/hooks/pre-push` (to create the hook file)

3. `nano .git/hooks/pre-push` (to edit the hook file)

4. Paste the following content in it and save:

```sh
#!/bin/bash

protected_branch='main'
current_branch=$(git symbolic-ref HEAD | sed -e 's,.*/\(.*\),\1,')

if [ $protected_branch = $current_branch ]
then
    echo "${protected_branch} is a protected branch, create PR to merge"
    exit 1 # push will not execute
else
    exit 0 # push will execute
fi
```
5. `chmod +x .git/hooks/pre-push` (to make the hook file executable)

Ref - https://hiltonmeyer.com/articles/protect-git-branch-and-prevent-master-push.html

**Comment by dineshram7660** [comment]
check this example

File: README.md [hunk 4]
```diff
@@ -7,20 +7,42 @@
...
 </p>
 
+
+## Documentation
+- in this project we are use admin panel for admin so admin can login and logout  and 
+- admin can add company brand name and edit brand name
+.
+
+## Important notes for developers
+- We are using OCI High availability MySQL instance in production. It requires that all the tables in the database must have primary keys.
```
Comment (dineshram7660) [inline README.md]: not requiring now plz remove this

File: README.md [hunk 5]
```diff
@@ -7,20 +7,42 @@
...
+
+## Technical
+- The information about the development process and reference details are recorded in <a href="https://github.com/jayesh-parmar/inventory-management/compare/Technical.md?expand=1" >this file</a>
```
Comment (dineshram7660) [inline README.md]: This file not now in over project so remove this

File: README.md (hunk 4 above)
Comment (dineshram7660) [inline README.md]: this not required now please remove this

**Review by dineshram7660 — CHANGES_REQUESTED** [review]

**Comment by dineshram7660** [comment]
hu tamne kai file examplae apu to ama hoy amj nai mukva nu pan apde jarur hoy avu j leva nu

**Comment by jayesh-parmar** [comment]
👍
